File: checks/ruby/lib/section508/checks/pointer_cancellation.rb
# frozen_string_literal: true

module Section508
  module Checks
    # The action must not run on down-event.
    class PointerCancellation
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        Dom.elements(root, "[ontouchstart], [onmousedown]").each do |element|
          handler = "#{element["ontouchstart"]} #{element["onmousedown"]}"
          if handler.gsub(/\s+/, "").length.positive?
            return ["“#{hint(element)}” completes on the down-event."]
          end
        end
        []
      end

      def self.hint(element)
        raw = element.text
        raw = nil if raw.empty?
        (raw || element.name.downcase).to_s.gsub(/\s+/, " ").strip
      end
    end
  end
end

File: checks/python/src/section508/checks/pointer_cancellation.py
from section508.dom import SPACE, attr, select, tag_name, text_of


def check(root):
    if not root:
        return []

    for element in select(root, "[ontouchstart], [onmousedown]"):
        handler = f"{attr(element, 'ontouchstart') or ''} {attr(element, 'onmousedown') or ''}"
        if SPACE.sub("", handler):
            return [f"“{hint(element)}” completes on the down-event."]
    return []


def hint(element) -> str:
    return SPACE.sub(" ", text_of(element) or tag_name(element)).strip()
